File: src/main/java/com/slimtrade/modules/theme/ThemeManager.java
package com.slimtrade.modules.theme;

import com.formdev.flatlaf.icons.FlatCheckBoxIcon;
import com.slimtrade.core.managers.FontManager;
import com.slimtrade.core.managers.SaveManager;
import com.slimtrade.core.utility.ZUtil;
import com.slimtrade.gui.buttons.IconButton;
import com.slimtrade.gui.windows.BasicDialog;
import com.slimtrade.modules.stopwatch.Stopwatch;
import com.slimtrade.modules.theme.components.ColorCheckbox;

import javax.imageio.ImageIO;
import javax.swing.*;
import javax.swing.plaf.BorderUIResource;
import javax.swing.plaf.FontUIResource;
import javax.swing.plaf.IconUIResource;
import java.awt.*;
import java.awt.datatransfer.StringSelection;
import java.awt.image.BufferedImage;
import java.awt.image.WritableRaster;
import java.io.File;
import java.io.IOException;
import java.util.List;
import java.util.*;

public class ThemeManager {

    // FIXME : Should this be moved to frame manager?
    private static final List<Component> frames = new ArrayList<>();

    // Listeners
    private static final ArrayList<IThemeListener> themeListeners = new ArrayList<>();
    private static final ArrayList<JComboBox<?>> stickyCombos = new ArrayList<>();
    private static final ArrayList<IFontChangeListener> fontChangeListeners = new ArrayList<>();
    private static final ArrayList<IDetailedFontChangeListener> detailedFontChangeListeners = new ArrayList<>();

    private static int cachedIconSize = 0;
    private static int currentFontSize;
    private static final HashMap<String, ImageIcon> iconMap = new HashMap<>();
    private static final HashMap<String, ImageIcon> colorIconMap = new HashMap<>();

    public static final Color TRANSPARENT = new Color(0, 0, 0, 0);
    public static final Color TRANSPARENT_CLICKABLE = new Color(0, 0, 0, 1);
    public static final Color POE_TEXT_DARK = new Color(53, 28, 13);
    public static final Color POE_TEXT_LIGHT = new Color(254, 192, 118);
    private static final int DEFAULT_OFFSET_COLOR_AMOUNT = 20;

    private static Theme currentTheme;
    private static String currentFontName;

    // Flags that monitor changes to font
    private static boolean iconSizeWasChanged = false;
    private static boolean fontSizeWasChanged = false;
    private static boolean fontStyleWasChanged = false;

    public static void addFrame(Component frame) {
        if (frames.contains(frame)) return;
        frames.add(frame);
    }

    public static void removeFrame(Component frame) {
        frames.remove(frame);
    }

    public static List<Component> getFrames() {
        return frames;
    }

    // Sticky combos are JComboBoxes that contain colored icons.
    // These combos need to have their selected values manually updated after switching themes
    public static void addStickyCombo(JComboBox<?> combo) {
        stickyCombos.add(combo);
    }

    public static void removeStickyCombo(JComboBox<?> combo) {
        stickyCombos.remove(combo);
    }

    public static void setTheme(Theme theme) {
        setTheme(theme, false);
    }

    public static void setTheme(Theme theme, boolean forceThemeRefresh) {
        assert (SwingUtilities.isEventDispatchThread());
        if (theme == null) theme = Theme.getDefaultColorTheme();
        if (theme == currentTheme && !forceThemeRefresh) return;
        int[] comboIcons = new int[stickyCombos.size()];
        for (int i = 0; i < stickyCombos.size(); i++) {
            comboIcons[i] = stickyCombos.get(i).getSelectedIndex();
        }
        iconMap.clear();
        colorIconMap.clear();
        currentTheme = theme;
        try {
            UIManager.setLookAndFeel(currentTheme.lookAndFeel);
        } catch (UnsupportedLookAndFeelException e) {
            e.printStackTrace();
        }
        patchTheme();
        for (Component frame : frames) {
            JRootPane rootPane = null;
            if (frame instanceof RootPaneContainer) rootPane = ((RootPaneContainer) frame).getRootPane();
            if (rootPane != null) {
                SwingUtilities.updateComponentTreeUI(rootPane);
                frame.revalidate();
                frame.repaint();
            }
        }
        for (int i = 0; i < stickyCombos.size(); i++) {
            stickyCombos.get(i).setSelectedIndex(comboIcons[i]);
        }
        for (IThemeListener listener : themeListeners) {
            listener.onThemeChange();
        }
    }

    public static Theme getCurrentTheme() {
        if (currentTheme == null) return Theme.getDefaultColorTheme();
        return currentTheme;
    }

    // Apply custom patching for theme issues. This is a bit hacky, but should be good enough until the themes are updated.
    private static void patchTheme() {
        // If the text areas and panels have the same background, add a border to the text area.
        Color panelBackground = UIManager.getColor("Panel.background");
        Color textAreaBackground = UIManager.getColor("TextArea.background");
        if (panelBackground.equals(textAreaBackground)) {
            Color buttonBorderColor = UIManager.getColor("Button.borderColor");
            UIManager.put("TextArea.border", new BorderUIResource(BorderFactory.createLineBorder(buttonBorderColor)));
        } else {
            UIManager.put("TextArea.border", new BorderUIResource(BorderFactory.createEmptyBorder()));
        }

        // Solarized Light Fix - Make checkboxes color match other inputs instead of being white
        if (currentTheme.name().equals("SOLARIZED_LIGHT")) {
            UIManager.put("CheckBox.icon", new IconUIResource(new ColorCheckbox()));
        } else {
            UIManager.put("CheckBox.icon", new IconUIResource(new FlatCheckBoxIcon()));
        }
    }

    // When getting an icon, size uses cached size, -1 uses unscaled size
    public static ImageIcon getIcon(String path) {
        return getIcon(path, cachedIconSize, true);
    }

    public static ImageIcon getIcon(String path, boolean resourceFolder) {
        return getIcon(path, cachedIconSize, resourceFolder);
    }

    public static ImageIcon getIcon(String path, int size) {
        return getIcon(path, size, true);
    }

    public static ImageIcon getIcon(String path, int size, boolean resourceFolder) {
        if (size == 0) size = cachedIconSize;
        // Return cached image if possible
        if (size == cachedIconSize && iconMap.containsKey(path)) {
            return iconMap.get(path);
        }
        try {
            BufferedImage img;
            if (resourceFolder) img = ImageIO.read(Objects.requireNonNull(ThemeManager.class.getResource(path)));
            else img = ImageIO.read(new File(path));
            if (img == null) return null; // This will only trigger with user submitted images
            ImageIcon icon;
            if (size == -1) {
                icon = new ImageIcon(img);
            } else {
                icon = new ImageIcon(img.getScaledInstance(size, size, Image.SCALE_SMOOTH));
            }
            if (size == cachedIconSize) iconMap.put(path, icon);
            return icon;
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
    }

    public static ImageIcon getColorIcon(String path) {
        return getColorIcon(path, cachedIconSize);
    }

    public static ImageIcon getColorIcon(String path, int size) {
        if (size == 0) size = cachedIconSize;
        // Return cached image if possible
        if (size == cachedIconSize && colorIconMap.containsKey(path)) {
            return colorIconMap.get(path);
        }
        // Generate new image
        try {
            BufferedImage img = ImageIO.read(Objects.requireNonNull(ThemeManager.class.getResource(path)));
            ImageIcon icon = new ImageIcon(getColorImage(img, UIManager.getColor("Button.foreground")).getScaledInstance(size, size, Image.SCALE_SMOOTH));
            if (size == cachedIconSize) colorIconMap.put(path, icon);
            return icon;
        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }

    private static Image getColorImage(BufferedImage image, Color color) {
        int width = image.getWidth();
        int height = image.getHeight();
        WritableRaster raster = image.getRaster();
        for (int xx = 0; xx < width; xx++) {
            for (int yy = 0; yy < height; yy++) {
                int[] pixels = raster.getPixel(xx, yy, (int[]) null);
                pixels[0] = color.getRed();
                pixels[1] = color.getGreen();
                pixels[2] = color.getBlue();
                raster.setPixel(xx, yy, pixels);
            }
        }
        return image;
    }

    public static void setFont(String fontName) {
        if (fontName == null) return;
        if (fontName.equals(currentFontName)) return;
        fontStyleWasChanged = true;
        Stopwatch.start();
        if (fontName.equals(currentFontName)) return;
        if (!FontManager.isValidFont(fontName)) return;
        FontManager.setPreferredFont(fontName);
        refreshDefaultFonts();
        for (Component component : frames) {
            changeComponentFont(component, FontManager.getPreferredFont());
        }
        currentFontName = fontName;
    }

    private static void refreshDefaultFonts() {
        if (FontManager.USE_SYSTEM_DEFAULT) return;
        Enumeration<Object> keys = UIManager.getDefaults().keys();
        String fontName = FontManager.getPreferredFont().getFontName();
        while (keys.hasMoreElements()) {
            Object key = keys.nextElement();
            Object value = UIManager.get(key);
            if (value instanceof FontUIResource) {
                Font oldFont = (FontUIResource) value;
                FontUIResource fontResource = new FontUIResource(fontName, oldFont.getStyle(), SaveManager.settingsSaveFile.data.fontSize);
                UIManager.put(key, fontResource);
            }
        }
    }

    private static void changeComponentFont(Component component, Font font) {
        Font previousFont = component.getFont();
        component.setFont(font.deriveFont(previousFont.getStyle(), previousFont.getSize()));
        if (component instanceof Container) {
            for (Component child : ((Container) component).getComponents()) {
                changeComponentFont(child, font);
            }
        }
    }

    public static void setFontSize(int size) {
        if (size == currentFontSize) return;
        currentFontSize = size;
        fontSizeWasChanged = true;
        refreshDefaultFonts();
        for (Component frame : frames) {
            setFontSizeRecursive(frame, size);
//            SwingUtilities.updateComponentTreeUI(frame);
            frame.revalidate();
            frame.repaint();
        }
    }

    private static void setFontSizeRecursive(Component component, int size) {
        if (component instanceof Container) {
            for (Component child : ((Container) component).getComponents()) {
                setFontSizeRecursive(child, size);
            }
        }
        Font curFont = component.getFont();
        component.setFont(curFont.deriveFont(curFont.getStyle(), size));
    }

    public static int getCachedIconSize() {
        return cachedIconSize;
    }

    public static void setIconSize(int size) {
        assert (SwingUtilities.isEventDispatchThread());
        if (size == cachedIconSize) return;
        iconSizeWasChanged = true;
        cachedIconSize = size;
        iconMap.clear();
        colorIconMap.clear();
        for (Component frame : frames) {
//            setIconSizeRecursive(frame, size);
            // FIXME : updateComponentThread is very inefficient, should create an alternative function
            SwingUtilities.updateComponentTreeUI(frame);
            frame.revalidate();
            frame.repaint();
            // FIXME
            if (frame instanceof BasicDialog) {
                ((BasicDialog) frame).pack();
            }
        }
    }

    private static void setIconSizeRecursive(Component component, int size) {
        if (component instanceof Container) {
            for (Component child : ((Container) component).getComponents()) {
                setIconSizeRecursive(child, size);
            }
        }
        if (component instanceof IconButton) {
            ((IconButton) component).setIconSize(size);
        }
    }

    public static void recursiveUpdateUI(JComponent component) {
        component.updateUI();
        for (Component child : component.getComponents()) {
            recursiveUpdateUI((JComponent) child);
        }
    }

    public static void checkFontChange() {
        boolean fontWasChanged = fontStyleWasChanged || fontSizeWasChanged || iconSizeWasChanged;
        if (!fontWasChanged) return;
        for (IFontChangeListener listener : fontChangeListeners) listener.onFontChanged();
        if (fontStyleWasChanged)
            for (IDetailedFontChangeListener listener : detailedFontChangeListeners) listener.onFontStyleChanged();
        if (fontSizeWasChanged)
            for (IDetailedFontChangeListener listener : detailedFontChangeListeners) listener.onFontSizeChanged();
        if (iconSizeWasChanged)
            for (IDetailedFontChangeListener listener : detailedFontChangeListeners) listener.onIconSizeChanged();
        fontStyleWasChanged = false;
        fontSizeWasChanged = false;
        iconSizeWasChanged = false;
    }

    //
    // Utility Color Functions
    //

    public static Color getDarkerColor(Color colorA, Color colorB) {
        if (colorIntValue(colorA) < colorIntValue(colorB)) return colorA;
        return colorB;
    }

    public static Color getLighterColor(Color colorA, Color colorB) {
        if (colorIntValue(colorA) > colorIntValue(colorB)) return colorA;
        return colorB;
    }

    public static Color modify(Color c, int mod) {
        int min = 0;
        int max = 255;
        int r = ZUtil.clamp(c.getRed() + mod, min, max);
        int g = ZUtil.clamp(c.getGreen() + mod, min, max);
        int b = ZUtil.clamp(c.getBlue() + mod, min, max);
        return new Color(r, g, b);
    }

    public static Color lighter(Color c) {
        return modify(c, DEFAULT_OFFSET_COLOR_AMOUNT);
    }

    public static Color lighter(Color c, int increase) {
        return modify(c, increase);
    }

    private static int colorIntValue(Color color) {
        return color.getRed() + color.getGreen() + color.getBlue();
    }

    public static Color adjustAlpha(Color color, int alpha) {
        return new Color(color.getRed(), color.getGreen(), color.getBlue(), alpha);
    }

    /**
     * Dumps all the Key, Value pairs from the UIManager into the clipboard.
     */
    public static void debugKeyValueDump() {
        Enumeration<Object> keys = UIManager.getDefaults().keys();
        StringBuilder builder = new StringBuilder();
        while (keys.hasMoreElements()) {
            Object key = keys.nextElement();
            Object value = UIManager.get(key);
            builder.append(key).append(" ::: ").append(value).append("\n");
        }
        StringSelection selection = new StringSelection(builder.toString());
        Toolkit.getDefaultToolkit().getSystemClipboard().setContents(selection, null);
    }

    // Listeners

    public static void addThemeListener(IThemeListener listener) {
        themeListeners.add(listener);
    }

    public static void removeThemeListener(IThemeListener listener) {
        themeListeners.remove(listener);
    }

    public static void addFontListener(IFontChangeListener listener) {
        fontChangeListeners.add(listener);
    }

    public static void removeFontChangeListener(IFontChangeListener listener) {
        fontChangeListeners.remove(listener);
    }

}

File: src/main/java/github/zmilla93/modules/theme/ThemeManager.kt
package github.zmilla93.modules.theme

import com.formdev.flatlaf.icons.FlatCheckBoxIcon
import github.zmilla93.core.managers.FontManager
import github.zmilla93.core.managers.SaveManager
import github.zmilla93.core.utility.ZUtil
import github.zmilla93.gui.buttons.IconButton
import github.zmilla93.gui.windows.BasicDialog
import github.zmilla93.modules.stopwatch.Stopwatch
import github.zmilla93.modules.theme.ThemeManager.frames
import github.zmilla93.modules.theme.components.ColorCheckbox
import github.zmilla93.modules.theme.extensions.ThemeExtension
import github.zmilla93.modules.theme.listeners.ColorblindChangeListener
import github.zmilla93.modules.theme.listeners.IDetailedFontChangeListener
import github.zmilla93.modules.theme.listeners.IFontChangeListener
import github.zmilla93.modules.theme.listeners.IThemeListener
import github.zmilla93.modules.updater.ZLogger
import io.github.zmilla93.modules.theme.UIProperty
import java.awt.*
import java.awt.datatransfer.StringSelection
import java.awt.image.BufferedImage
import java.io.File
import java.io.IOException
import java.net.URL
import java.util.*
import javax.imageio.ImageIO
import javax.swing.*
import javax.swing.plaf.BorderUIResource
import javax.swing.plaf.FontUIResource
import javax.swing.plaf.IconUIResource

object ThemeManager {
    // FIXME : Should this be moved to frame manager?
    val frames: MutableList<Component> = ArrayList<Component>()

    // Listeners
    private val themeListeners = ArrayList<IThemeListener>()
    private val stickyCombos = ArrayList<JComboBox<*>?>()
    private val fontChangeListeners = ArrayList<IFontChangeListener>()
    private val detailedFontChangeListeners = ArrayList<IDetailedFontChangeListener>()
    private val colorblindChangeListeners = ArrayList<ColorblindChangeListener>()

    var cachedIconSize: Int = 18
        private set
    private var currentFontSize = 0
    private val iconMap = HashMap<String?, ImageIcon?>()
    private val colorIconMap = HashMap<String?, ImageIcon?>()

    @JvmField
    val TRANSPARENT: Color = Color(0, 0, 0, 0)

    @JvmField
    val TRANSPARENT_CLICKABLE: Color = Color(0, 0, 0, 1)

    @JvmField
    val POE_TEXT_DARK: Color = Color(53, 28, 13)

    @JvmField
    val POE_TEXT_LIGHT: Color = Color(254, 192, 118)
    private const val DEFAULT_OFFSET_COLOR_AMOUNT = 20

    private var currentTheme: Theme? = null
    private var currentFontName: String? = null

    @JvmStatic
    var isColorblindMode: Boolean = false
        set(state) {
            if (state == field) return
            field = state
            UIManager.setLookAndFeel(currentTheme!!.lookAndFeel.javaClass.newInstance())
            updateAllComponentTrees()
            for (listener in colorblindChangeListeners) listener.onColorblindChange(state)
        }

    // Flags that monitor changes to font
    private var iconSizeWasChanged = false
    private var fontSizeWasChanged = false
    private var fontStyleWasChanged = false

    @JvmStatic
    fun addFrame(frame: Component?) {
        if (frames.contains(frame)) return
        frames.add(frame!!)
    }

    @JvmStatic
    fun removeFrame(frame: Component?) {
        frames.remove(frame)
    }

    // Sticky combos are JComboBoxes that contain colored icons.
    // These combos need to have their selected values manually updated after switching themes
    @JvmStatic
    fun addStickyCombo(combo: JComboBox<*>?) {
        stickyCombos.add(combo)
    }

    @JvmStatic
    fun removeStickyCombo(combo: JComboBox<*>?) {
        stickyCombos.remove(combo)
    }

    @JvmStatic
    fun setTheme(theme: Theme?) {
        setTheme(theme, false)
    }

    @JvmStatic
    fun setTheme(theme: Theme?, forceThemeRefresh: Boolean) {
        var theme = theme
        assert(SwingUtilities.isEventDispatchThread())
        if (theme == null) theme = Theme.getDefaultColorTheme()
        if (theme == currentTheme && !forceThemeRefresh) return
        val comboIcons = IntArray(stickyCombos.size)
        for (i in stickyCombos.indices) {
            comboIcons[i] = stickyCombos.get(i)!!.getSelectedIndex()
        }
        iconMap.clear()
        colorIconMap.clear()
        currentTheme = theme
        try {
            UIManager.setLookAndFeel(currentTheme!!.lookAndFeel.javaClass.newInstance())
        } catch (e: UnsupportedLookAndFeelException) {
            throw RuntimeException("Unsupported theme!")
        } catch (e: IllegalAccessException) {
            throw RuntimeException("Unsupported theme!")
        } catch (e: InstantiationException) {
            throw RuntimeException("Unsupported theme!")
        }
        patchTheme()
        updateAllComponentTrees()
        for (i in stickyCombos.indices) {
            stickyCombos.get(i)!!.setSelectedIndex(comboIcons[i])
        }
        for (listener in themeListeners) {
            listener.onThemeChange()
        }
    }

    /** Call this anytime [UIProperty]s need to be reevaluated or reapplied. */
    fun updateUIProperties() {
        updateAllComponents {
            // FIXME @important : Check if default font size is enabled
            // Apply default font size
            it.font = it.font.deriveFont(FontManager.fontSize.toFloat())
            for (prop in UIProperty.entries) {
                val value = UIProperty.getProperty<Any>(it, prop)
                if (value != null) {
                    // Apply UI Property
                    prop.runner.applyProperty(it, value)
                }
            }
        }
    }

    /** Runs an update function on every component in [frames] window hierarchy. */
    fun updateAllComponents(updater: ComponentUpdater) {
        for (frame in frames) {
            var root: Container? = null
            if (frame is JFrame) root = frame.contentPane
            else if (frame is JDialog) root = frame.contentPane
            if (root is JComponent) updateComponentRecursive(root, updater)
        }
    }

    /**
     * Runs an updater function [ComponentUpdater] on a component,
     * as well as any child components.
     */
    private fun updateComponentRecursive(comp: JComponent, updater: ComponentUpdater) {
        updater.update(comp)
        comp.components.forEach { if (it is JComponent) updateComponentRecursive(it, updater) }
    }

    fun setThemeFastDebug(theme: Theme?) {
        setThemeFastDebug(theme, true)
    }

    @JvmStatic
    fun setThemeFastDebug(theme: Theme?, rebuildIcons: Boolean) {
        currentTheme = theme
        if (rebuildIcons) colorIconMap.clear()
        try {
            UIManager.setLookAndFeel(currentTheme!!.lookAndFeel)
        } catch (e: UnsupportedLookAndFeelException) {
            throw RuntimeException("Unsupported theme!")
        }
    }

    private fun updateAllComponentTrees() {
        for (frame in frames) SwingUtilities.updateComponentTreeUI(frame)
    }


    @JvmStatic
    fun getCurrentTheme(): Theme {
        if (currentTheme == null) return Theme.getDefaultColorTheme()
        return currentTheme!!
    }

    @JvmStatic
    fun extensions(): ThemeExtension? {
        return getCurrentTheme().extensions
    }

    // FIXME: Review this
    // Apply custom patching for theme issues. This is a bit hacky, but should be good enough until the themes are updated.
    private fun patchTheme() {
        // If the text areas and panels have the same background, add a border to the text area.
        val panelBackground = UIManager.getColor("Panel.background")
        val textAreaBackground = UIManager.getColor("TextArea.background")
        if (panelBackground == textAreaBackground) {
            val buttonBorderColor = UIManager.getColor("Button.borderColor")
            UIManager.put("TextArea.border", BorderUIResource(BorderFactory.createLineBorder(buttonBorderColor)))
        } else {
            UIManager.put("TextArea.border", BorderUIResource(BorderFactory.createEmptyBorder()))
        }

        // Solarized Light Fix - Make checkboxes color match other inputs instead of being white
        if (currentTheme!!.name == "SOLARIZED_LIGHT") {
            UIManager.put("CheckBox.icon", IconUIResource(ColorCheckbox()))
        } else {
            UIManager.put("CheckBox.icon", IconUIResource(FlatCheckBoxIcon()))
        }
    }

    // When getting an icon, size uses cached size, -1 uses unscaled size
    @JvmStatic
    fun getIcon(path: String): ImageIcon? {
        return getIcon(path, cachedIconSize, true)
    }

    fun getIcon(path: String, resourceFolder: Boolean): ImageIcon? {
        return getIcon(path, cachedIconSize, resourceFolder)
    }

    @JvmStatic
    fun getIcon(path: String, size: Int): ImageIcon? {
        return getIcon(path, size, true)
    }

    @JvmStatic
    fun getIcon(path: String, size: Int, resourceFolder: Boolean): ImageIcon? {
        var size = size
        if (size == 0) size = cachedIconSize
        // Return cached image if possible
        if (size == cachedIconSize && iconMap.containsKey(path)) {
            return iconMap.get(path)
        }
        try {
            val img: BufferedImage?
            if (resourceFolder) img =
                ImageIO.read(Objects.requireNonNull<URL?>(ThemeManager::class.java.getResource(path)))
            else img = ImageIO.read(File(path))
            if (img == null) return null // This will only trigger with user submitted images

            var width = size
            var height = size
            // Scale non square images
            if (img.getWidth() > img.getHeight()) {
                val mult = size.toFloat() / img.getHeight()
                width = Math.round(img.getWidth() * mult)
            } else {
                val mult = size.toFloat() / img.getWidth()
                height = Math.round(img.getHeight() * mult)
            }
            val icon: ImageIcon?
            if (size == -1) {
                icon = ImageIcon(img)
            } else {
                icon = ImageIcon(img.getScaledInstance(width, height, Image.SCALE_SMOOTH))
            }
            if (size == cachedIconSize) iconMap.put(path, icon)
            return icon
        } catch (e: IOException) {
            ZLogger.err("Could not find image: " + path)
            ZLogger.err(e.getStackTrace())
            return null
        } catch (e: NullPointerException) {
            ZLogger.err("Could not find image: " + path)
            ZLogger.err(e.getStackTrace())
            return null
        }
    }

    @JvmStatic
    fun getColorIcon(path: String): ImageIcon? {
        return getColorIcon(path, cachedIconSize)
    }

    @JvmStatic
    fun getColorIcon(path: String, size: Int): ImageIcon? {
        var size = size
        if (size == 0) size = cachedIconSize
        // Return cached image if possible
        if (size == cachedIconSize && colorIconMap.containsKey(path)) {
            return colorIconMap.get(path)
        }
        // Generate new image
        try {
            val img = ImageIO.read(Objects.requireNonNull<URL?>(ThemeManager::class.java.getResource(path)))
            val icon = ImageIcon(
                getColorImage(img, UIManager.getColor("Button.foreground")).getScaledInstance(
                    size,
                    size,
                    Image.SCALE_SMOOTH
                )
            )
            if (size == cachedIconSize) colorIconMap.put(path, icon)
            return icon
        } catch (e: IOException) {
            ZLogger.err("Invalid path: " + path)
            ZLogger.err(e.getStackTrace())
        } catch (e: NullPointerException) {
            ZLogger.err("Invalid path: " + path)
            ZLogger.err(e.getStackTrace())
        }
        return null
    }

    private fun getColorImage(image: BufferedImage, color: Color): Image {
        val width = image.getWidth()
        val height = image.getHeight()
        val raster = image.getRaster()
        for (xx in 0..<width) {
            for (yy in 0..<height) {
                val pixels = raster.getPixel(xx, yy, null as IntArray?)
                pixels[0] = color.getRed()
                pixels[1] = color.getGreen()
                pixels[2] = color.getBlue()
                raster.setPixel(xx, yy, pixels)
            }
        }
        return image
    }

    @JvmStatic
    fun setFont(fontName: String?) {
        if (fontName == null) return
        if (fontName == currentFontName) return
        fontStyleWasChanged = true
        Stopwatch.start()
        if (fontName == currentFontName) return
        if (!FontManager.isValidFont(fontName)) return
        FontManager.setPreferredFont(fontName)
        refreshDefaultFonts()
        for (component in frames) {
            changeComponentFont(component, FontManager.getPreferredFont())
        }
        currentFontName = fontName
    }

    private fun refreshDefaultFonts() {
        if (FontManager.USE_SYSTEM_DEFAULT) return
        val keys = UIManager.getDefaults().keys()
        val fontName = FontManager.getPreferredFont().getFontName()
        while (keys.hasMoreElements()) {
            val key = keys.nextElement()
            val value = UIManager.get(key)
            if (value is FontUIResource) {
                val oldFont: Font = value
                val fontResource =
                    FontUIResource(fontName, oldFont.getStyle(), SaveManager.settingsSaveFile.data.fontSize)
                UIManager.put(key, fontResource)
            }
        }
    }

    private fun changeComponentFont(component: Component, font: Font) {
        val previousFont = component.getFont()
        component.setFont(font.deriveFont(previousFont.getStyle(), previousFont.getSize().toFloat()))
        if (component is Container) {
            for (child in component.getComponents()) {
                changeComponentFont(child, font)
            }
        }
    }

    @JvmStatic
    fun setFontSize(size: Int) {
        if (size == FontManager.fontSize) return
        return
        FontManager.fontSize = size
        fontSizeWasChanged = true
        // FIXME : listeners?
//        updateUIProperties()

//        refreshDefaultFonts()
//        for (frame in frames) {
//            setFontSizeRecursive(frame, size)
//            //            SwingUtilities.updateComponentTreeUI(frame);
//            frame.revalidate()
//            frame.repaint()
//        }
    }

    private fun setFontSizeRecursive(component: Component, size: Int) {
        if (component is Container) {
            for (child in component.getComponents()) {
                setFontSizeRecursive(child, size)
            }
        }
        val curFont = component.getFont()
        component.setFont(curFont.deriveFont(curFont.getStyle(), size.toFloat()))
    }

    @JvmStatic
    fun setIconSize(size: Int) {
        assert(SwingUtilities.isEventDispatchThread())
        if (size == cachedIconSize) return
        iconSizeWasChanged = true
        cachedIconSize = size
        iconMap.clear()
        colorIconMap.clear()
        for (frame in frames) {
//            setIconSizeRecursive(frame, size);
            // FIXME : updateComponentThread is very inefficient, should create an alternative function
            SwingUtilities.updateComponentTreeUI(frame)
            frame.revalidate()
            frame.repaint()
            // FIXME
            if (frame is BasicDialog) {
                frame.pack()
            }
        }
    }

    private fun setIconSizeRecursive(component: Component?, size: Int) {
        if (component is Container) {
            for (child in component.getComponents()) {
                setIconSizeRecursive(child, size)
            }
        }
        if (component is IconButton) {
            component.setIconSize(size)
        }
    }

    fun recursiveUpdateUI(component: JComponent) {
        component.updateUI()
        for (child in component.getComponents()) {
            ThemeManager.recursiveUpdateUI((child as JComponent?)!!)
        }
    }

    @JvmStatic
    fun checkFontChange() {
        val fontWasChanged = fontStyleWasChanged || fontSizeWasChanged || iconSizeWasChanged
        if (!fontWasChanged) return
        for (listener in fontChangeListeners) listener.onFontChanged()
        if (fontStyleWasChanged) for (listener in detailedFontChangeListeners) listener.onFontStyleChanged()
        if (fontSizeWasChanged) for (listener in detailedFontChangeListeners) listener.onFontSizeChanged()
        if (iconSizeWasChanged) for (listener in detailedFontChangeListeners) listener.onIconSizeChanged()
        fontStyleWasChanged = false
        fontSizeWasChanged = false
        iconSizeWasChanged = false
    }

    //
    // Utility Color Functions
    //
    fun getDarkerColor(colorA: Color, colorB: Color): Color {
        if (colorIntValue(colorA) < colorIntValue(colorB)) return colorA
        return colorB
    }

    @JvmStatic
    fun getLighterColor(colorA: Color, colorB: Color): Color {
        if (colorIntValue(colorA) > colorIntValue(colorB)) return colorA
        return colorB
    }

    fun getBrightness(color: Color): Float {
        return 0.2126f * color.getRed() + 0.7152f * color.getGreen() + 0.0722f * color.getBlue()
    }

    fun isDark(color: Color): Boolean {
        return getBrightness(color) < 128
    }

    fun modify(c: Color, mod: Int): Color {
        val min = 0
        val max = 255
        val r = ZUtil.clamp(c.getRed() + mod, min, max)
        val g = ZUtil.clamp(c.getGreen() + mod, min, max)
        val b = ZUtil.clamp(c.getBlue() + mod, min, max)
        return Color(r, g, b)
    }

    @JvmStatic
    fun lighter(c: Color): Color {
        return modify(c, DEFAULT_OFFSET_COLOR_AMOUNT)
    }

    fun lighter(c: Color, increase: Int): Color {
        return modify(c, increase)
    }

    private fun colorIntValue(color: Color): Int {
        return color.getRed() + color.getGreen() + color.getBlue()
    }

    @JvmStatic
    fun adjustAlpha(color: Color, alpha: Int): Color {
        return Color(color.getRed(), color.getGreen(), color.getBlue(), alpha)
    }

    /**
     * Dumps all the Key, Value pairs from the UIManager into the clipboard.
     */
    @JvmStatic
    fun debugKeyValueDump() {
        val keys = UIManager.getDefaults().keys()
        val builder = StringBuilder()
        while (keys.hasMoreElements()) {
            val key = keys.nextElement()
            val value = UIManager.get(key)
            builder.append(key).append(" ::: ").append(value).append("\n")
        }
        val selection = StringSelection(builder.toString())
        Toolkit.getDefaultToolkit().getSystemClipboard().setContents(selection, null)
    }

    // Listeners
    @JvmStatic
    fun addThemeListener(listener: IThemeListener?) {
        themeListeners.add(listener!!)
    }

    @JvmStatic
    fun removeThemeListener(listener: IThemeListener?) {
        themeListeners.remove(listener)
    }

    // FIXME : Font listener should be moved to FontManager
    @JvmStatic
    fun addFontListener(listener: IFontChangeListener?) {
        fontChangeListeners.add(listener!!)
    }

    @JvmStatic
    fun removeFontChangeListener(listener: IFontChangeListener?) {
        fontChangeListeners.remove(listener)
    }

    @JvmStatic
    fun addColorblindChangeListener(listener: ColorblindChangeListener?) {
        colorblindChangeListeners.add(listener!!)
    }

    fun removeColorblindChangeListener(listener: ColorblindChangeListener?) {
        colorblindChangeListeners.remove(listener)
    }
}
